sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgObjectFactory.cs

```diff
@@ -0,0 +1,147 @@
+
+namespace SonarAnalyzer.ControlFlowGraph.CSharp
+{
+    public class UcfgObjectFactory
+    {
+        /// <summary>
+        /// The string constant representation in the Sonar Security engine (Java part). When
+        /// an instruction receives or returns a type that is not string we use this instead
+        /// of a variable.
+        /// </summary>
+        private static readonly Expression ConstantExpression = new Expression
+        {
+            Const = new Constant { Value = "\"\"" }
+        };
+
+        private readonly Dictionary<SyntaxNode, Expression> nodeExpressionMap = new Dictionary<SyntaxNode, Expression>();
+
+        private int tempVariablesCounter;
+
+        public UCFG CreateUcfg(SyntaxNode syntaxNode, UcfgMethod method) =>
+            new UCFG
+            {
+                Location = GetLocation(syntaxNode),
+                MethodId = method.ToString(),
+            };
+
+        public BasicBlock CreateUcfgBlock(string id) =>
+            new BasicBlock { Id = id };
+
+        public Instruction CreateMethodCall(SyntaxNode syntaxNode, UcfgMethod method, params SyntaxNode[] arguments) =>
+            CreateInstruction(syntaxNode, method, CreateTempVariable(), arguments);
+
+        public Instruction CreateConcatenation(SyntaxNode syntaxNode, params SyntaxNode[] arguments) =>
+            CreateInstruction(syntaxNode, UcfgMethod.Concatenation, CreateTempVariable(), arguments);
+
+        public Instruction CreateAssignment(SyntaxNode syntaxNode, string variableName, SyntaxNode argument) =>
+            CreateInstruction(syntaxNode, UcfgMethod.Assignment, variableName, argument);
+
+        public Instruction CreateEntryPoint(SyntaxNode syntaxNode, params SyntaxNode[] arguments) =>
+            CreateInstruction(syntaxNode, UcfgMethod.EntryPoint, CreateTempVariable(), arguments);
+
+        public Instruction CreateParameterAnnotation(string parameterName, SyntaxNode attributeSyntax) =>
+            CreateInstruction(attributeSyntax, UcfgMethod.Annotation, parameterName, attributeSyntax);
+
+        public Instruction CreateVariable(SyntaxNode syntaxNode, string variableName)
+        {
+            nodeExpressionMap[syntaxNode] = CreateVariableExpression(variableName);
+            return null;
+        }
+
+        public Instruction CreateConstant(SyntaxNode syntaxNode)
+        {
+            nodeExpressionMap[syntaxNode] = ConstantExpression;
+            return null;
+        }
+
+        public Return CreateReturnExpression(SyntaxNode syntaxNode = null, SyntaxNode returnedValue = null) =>
+            new Return
+            {
+                Location = syntaxNode == null
+                    ? null
+                    : GetLocation(syntaxNode),
+                ReturnedExpression = returnedValue == null
+                    ? ConstantExpression
+                    : GetMappedExpression(returnedValue),
+            };
+
+        public Jump CreateJump(params string[] blockIds)
```

Comment:
As far as I understand this should belong to the `UcfgBuilder` class.

Reply:
Nope. UcfgBuilder generates UCFG CFG; UcfgObjectFactory generates UCFG objects (the protobuf generated classes) from SyntaxNodes.

Reply:
Jump is a block element? Is it some instruction?